### PR TITLE
Ray-based Tuner for LabelModel

### PR DESCRIPTION
**Pull request by paroma**
## Description of proposed changes

- Add LabelModelTuner that uses [Ray Tune](https://ray.readthedocs.io/en/latest/tune.html) to search over specified LabelModel numerical hyper parameters
     - Add ray to requirements

## Test plan
- Add tests for random search and grid search
- Add complex test with Ray Hyperband scheduler with random search

## Checklist

Need help on these? Just ask!

* [x] I have read the **CONTRIBUTING** document.
* [x] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.
* [ ] I have run `tox -e complex` and/or `tox -e spark` if appropriate.
* [x] All new and existing tests passed.

### CODE REVIEW

**Comment by ajratner** [comment]
@paroma from quick skim: LGTM!!  But I wonder: would it be worth thinking about some heuristics we might consider for when `Y_dev` is not available?  Maybe something for a separate commit, and just get this one in first, but can briefly discuss tomorrow?

File: snorkel/labeling/__init__.py [hunk 1]
```diff
@@ -10,4 +10,5 @@
     RandomVoter,
 )
 from .model.label_model import LabelModel  # noqa: F401
+from .model.tuner import LabelModelTuner  # noqa: F401
```
Comment (henryre) [inline snorkel/labeling/__init__.py]: need to remove to keep `ray` off the main import path. can revisit later, but doesn't look essential right now.

File: snorkel/labeling/model/tuner.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        Gold labels for dev set for tuning hyperparameters
+    metric
+        Metric to maximize while hyperparameter tuning
+    fit_config
+        Arguments for changing LabelModel fit() config defaults
+    """
+
+    def __init__(
+        self,
+        label_model,
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: type plz!

File: snorkel/labeling/model/tuner.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        L_train: np.ndarray,
+        L_dev: np.ndarray,
+        Y_dev: np.ndarray,
+        metric: str = "accuracy",
+        fit_config: Optional[Dict[str, Any]] = {},
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: No mutable defaults, just switch to `None`

File: snorkel/labeling/model/tuner.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    ) -> None:
+
+        self.label_model = label_model
+        self.L_train = L_train
+        self.L_dev = L_dev
+        self.Y_dev = Y_dev
+        self.metric = metric
+        self.fit_config = fit_config
+
+    def _trainable(self):  # type: ignore
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: `def _trainable(self) -> "LabelModelTrainable":` or `def _trainable(self) -> Trainable:`

File: snorkel/labeling/model/tuner.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        """Return a Trainable class definition for Ray tuner with some user-defined parameters from LabelModelTuner class."""
+
+        class LabelModelTrainable(Trainable):
+            """Trainable class for LabelModel for Ray tuner."""
+
+            def _setup(self, config: Optional[Dict[str, Any]]):
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: return types for all class methods

File: snorkel/labeling/model/tuner.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+                self.tuner = LabelModelTrainable.tuner  # type: ignore
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: I'm guessing this is some ray pattern, but i don't see why this is necessary

File: snorkel/labeling/model/tuner.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+            def _save(self, checkpoint_dir: str):
+                path = os.path.join(checkpoint_dir, "checkpoint")
+                with open(path, "w") as f:
+                    f.write(json.dumps({"timestep": self.timestep}))
+                return path
+
+            def _restore(self, checkpoint_path: str):
+                with open(checkpoint_path) as f:
+                    self.timestep = json.loads(f.read())["timestep"]
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: `json.load(f)`

Comment (henryre) [reply snorkel/labeling/model/tuner.py]: what about the config?

File: snorkel/labeling/model/tuner.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        return LabelModelTrainable
+
+    def run_tuner(
+        self,
+        config: Dict[str, Any],
+        scheduler: Optional[Any] = None,
+        name: Optional[str] = "label_model_tuner",
+        num_samples: Optional[int] = 5,
+        verbose: Optional[bool] = True,
+        experiment_configs: Optional[Dict[str, Any]] = {
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: no mutable defaults

File: snorkel/labeling/model/tuner.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+            "stop": {"training_iteration": 100}
+        },
+        run_configs: Optional[Dict[str, Any]] = {
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: no mutable defaults

File: snorkel/labeling/model/tuner.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+            def _restore(self, checkpoint_path: str):
+                with open(checkpoint_path) as f:
+                    self.timestep = json.loads(f.read())["timestep"]
+
+        # returning a LabelModelTrainable class definition with tuner as user-defined LabelModelTuner instance
+        LabelModelTrainable.tuner = self
+        return LabelModelTrainable
+
+    def run_tuner(
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: `tune`?

File: snorkel/labeling/model/tuner.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+        self,
+        config: Dict[str, Any],
+        scheduler: Optional[Any] = None,
+        name: Optional[str] = "label_model_tuner",
+        num_samples: Optional[int] = 5,
+        verbose: Optional[bool] = True,
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: not used

File: snorkel/labeling/model/tuner.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+        self.L_train = L_train
+        self.L_dev = L_dev
+        self.Y_dev = Y_dev
+        self.metric = metric
+        self.fit_config = fit_config
+
+    def _trainable(self):  # type: ignore
+        """Return a Trainable class definition for Ray tuner with some user-defined parameters from LabelModelTuner class."""
+
+        class LabelModelTrainable(Trainable):
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: why do we need this to be defined here?

File: snorkel/labeling/model/tuner.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+            config=config,
+            **experiment_configs,
+        )
+
+        analysis = ray.tune.run(exp, scheduler=scheduler, **run_configs)
+        self.best_config = analysis.get_best_config(metric=self.metric)
+        self.label_model.fit(
+            self.L_train,
+            **self.fit_config,
+            **analysis.get_best_config(metric=self.metric),
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: `**self.best_config`

File: snorkel/labeling/model/tuner.py [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    """A hyperparameter tuner for the LabelModel using Ray Tune (https://ray.readthedocs.io/en/latest/tune.html).
+
+    Example
+    -------
+    >>> L_train = np.array([[1, 1, 1, 0], [0, 1, 1, 0], [1, 0, 0, 1], [0, 1, 0, 1]])
+    >>> L_dev = np.array([[1, 1, 1, 0], [0, 1, 1, 0], [1, 0, 0, 1], [0, 1, 0, 1]])
+    >>> Y_dev = np.array([1, 1, 0, 1])
+    >>> label_model = LabelModel(cardinality=2)
+    >>> label_model_tuner = LabelModelTuner(label_model, L_train, L_dev, Y_dev)
+    >>> label_model_tuner = LabelModelTuner(label_model, L_train, L_dev, Y_dev, fit_config={"class_balance" : np.array([0.5, 0.5])})
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: nit: line is too long

File: snorkel/labeling/model/tuner.py [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+        Parameters
+        ----------
+        config
+            Dictionary of hyperparameters for LabelModel fit() to search over
+        scheduler
+            Special scheduler for hyperparameter search trials
+        name
+            Name of experiment
+        num_samples
+            Number of trials to run. Set to 1 for gridsearch
```
Comment (henryre) [inline snorkel/labeling/model/tuner.py]: From my reading of this, 1 would mean pick one random set of hyperparams and just run that (i.e. not a search). is it a flag value?

File: snorkel/labeling/model/tuner.py (hunk 6 above)
Comment (paroma) [reply snorkel/labeling/model/tuner.py]: Ray requires a Trainable with specific forms for the _train(), _setup() methods, but doesn't have a clean way of taking in additional args for those methods. This links to the overall LabelModelTuner class when Trainable is defined so it has access to the args that are constant across all runs and required across fit/score etc.

File: snorkel/labeling/model/tuner.py (hunk 7 above)
Comment (paroma) [reply snorkel/labeling/model/tuner.py]: This is from the hyberband example [here](https://github.com/ray-project/ray/blob/9ce6dd9b883594cec4ce42bd55962d53088afabe/python/ray/tune/examples/hyperband_example.py#L38) 

On second read, the _setup function there only has timestep while I have others, will fix here

File: snorkel/labeling/model/tuner.py (hunk 12 above)
Comment (paroma) [reply snorkel/labeling/model/tuner.py]: https://github.com/snorkel-team/snorkel/pull/1457#discussion_r323348959

it refers to the LabelModelTuner class that it's defined in. Ray requires a class object when the tuner is called, and this returns a custom class object every time

File: snorkel/labeling/model/tuner.py (hunk 15 above)
Comment (paroma) [reply snorkel/labeling/model/tuner.py]: There's no flag value iirc, but with grid search, one "sample" goes through all the possible combinations by default.

I can change default for num_smaples to be 1 and then set it explicitly for the rest

**Comment by paroma** [comment]
@ajratner agreed - we can have something without Y_dev that depends on looking at the learned weights or patterns in assigned labels. should be easy to edit the existing LabelModelTuner class to check if Y_dev exists!

**Comment by paroma** [comment]
Closing in favor of #1461, discussion related to hyper parameter search without Y_dev still applies
